### PR TITLE
[MRG+1] MaxNLocator 'steps' validation; and documentation. Closes #7578.

### LINKED ISSUE #7578: Validate steps input to `MaxNLocator`
**NOTE** The following was actually incorrect as the `steps` input was invalid; thus this ticket has been modified slightly to note the fact that `steps` should be validated in some manner.

Differences in `MaxNLocator` between 2.0.0 and 1.5.3 may be minimized by ensuring `steps` is between 1 and 10 and the classic style is used.

---

I ran the Cartopy test suite against v2.0.0rc1 and there appear to be some changes in the result from `MaxNLocator`. The default style change page only references changes to `AutoLocator` and the API changes page mentions `LinearScaleLocator`. I don't see any obvious changes in the tests for `MaxNLocator`, so I guess this is a bug, or needs to be better documented.

Here is the result on `1.5.3`:
```python
$ python
>>> import matplotlib
>>> matplotlib.__version__
'1.5.3'
>>> from matplotlib.ticker import MaxNLocator
>>> degree_locator = MaxNLocator(nbins=9, steps=[1, 2, 3, 6, 15, 18])
>>> degree_locator.tick_values(-180, 180)
array([-180., -120.,  -60.,    0.,   60.,  120.,  180.])
>>> degree_locator.tick_values(-50000000.0, 50000000.0)
array([-60000000., -40000000., -20000000.,         0.,  20000000.,
        40000000.,  60000000.])
>>> degree_locator.tick_values(-61.2479231, 68.41835367)
array([-80., -60., -40., -20.,   0.,  20.,  40.,  60.,  80.])
```
and here is the result with `v2.0.0rc1`
```python
$ python
>>> import matplotlib
>>> matplotlib.__version__
'2.0.0rc1'
>>> from matplotlib.ticker import MaxNLocator
>>> degree_locator = MaxNLocator(nbins=9, steps=[1, 2, 3, 6, 15, 18])
>>> degree_locator.tick_values(-180, 180)
array([-180., -120.,  -60.,    0.,   60.,  120.,  180.])
>>> degree_locator.tick_values(-50000000.0, 50000000.0)
array([-60000000., -45000000., -30000000., -15000000.,         0.,
        15000000.,  30000000.,  45000000.,  60000000.])
>>> degree_locator.tick_values(-61.2479231, 68.41835367)
array([-75., -60., -45., -30., -15.,   0.,  15.,  30.,  45.,  60.,  75.])
```

- [x] Matplotlib version, Python version and Platform (Windows, OSX, Linux ...): 2.0.0rc1, conda's 3.5.2, Linux
- [x] How did you install Matplotlib and Python (pip, anaconda, from source ...): from source
- [x] If possible please supply a [Short, Self Contained, Correct, Example](http://sscce.org/)
      that demonstrates the issue i.e a small piece of code which reproduces the issue
      and can be run with out any other (or as few as possible) external dependencies.
- [ ] If this is an image generation bug attach a screenshot demonstrating the issue.
- [x] If this is a regression (Used to work in an earlier version of Matplotlib), please 
      note where it used to work.

### PR DESCRIPTION
**Pull request by efiring**
This adds validation of the "steps" kwarg, fixes a bug in the implementation of the "integer" kwarg, and adds a note to dflt_style_changes.rst regarding the algorithm change in MaxNLocator.

### CODE REVIEW

File: lib/matplotlib/ticker.py [hunk 1]
```diff
@@ -1685,6 +1685,33 @@ def __init__(self, *args, **kwargs):
...
+        if steps[-1] != 10:
+            steps = np.hstack((steps, 10))
+        return steps
+
+    @staticmethod
+    def _staircase(steps):
+        # Make an extended staircase within which the needed
+        # step will be found.  This is probably much larger
+        # than necessary.
+        flights = (0.1 * steps[:-1], steps, 10 * steps[1])
```
Comment (QuLogic) [inline lib/matplotlib/ticker.py]: OK, I never noticed this on the previous PR, but why is the third entry ten times the second step, and not ten times the second-to-the-last steps in the same way as the first element?

File: lib/matplotlib/ticker.py [hunk 2]
```diff
@@ -1685,6 +1685,33 @@ def __init__(self, *args, **kwargs):
...
 
+    @staticmethod
+    def _validate_steps(steps):
+        if not np.iterable(steps):
+            raise ValueError('steps argument must be a sequence of numbers '
+                             'from 1 to 10')
+        steps = np.asarray(steps)
+        if np.any(np.diff(steps) <= 0):
+            raise ValueError('steps argument must be uniformly increasing')
+        if np.any((steps > 10) | (steps < 1)):
```
Comment (QuLogic) [inline lib/matplotlib/ticker.py]: Since you've already checked that it's increasing, you only need to check the first/last elements, I think.

Comment (efiring) [inline lib/matplotlib/ticker.py]: Good point.

File: lib/matplotlib/ticker.py (hunk 1 above)
Comment (QuLogic) [inline lib/matplotlib/ticker.py]: Right, so I was wondering why 20 and not all the way to 100? Is it just redundant?

Comment (efiring) [inline lib/matplotlib/ticker.py]: The strategy is to start with the largest step and work down to smaller ones, finding the smallest step that doesn't lead to too many intervals.  Everything is first scaled by a power of 10 based on an initial estimate.  Normally, this means that starting at the 10 would be adequate, but occasionally one step larger is needed, hence the inclusion of the next step past 10.  I couldn't see any circumstance that would require starting with a larger step than that, so I stopped there.

**Pull request merged**